File: src/eviction_parallel_handler/eviction_parallel_handler_workflow.erl
-module(eviction_parallel_handler_workflow).

% elp:ignore W0012 W0040
-moduledoc false.

-export([
    execute/2,
    eviction_handler/2
]).

-include_lib("temporal_sdk/include/workflow.hrl").

-define(ACTIVITY_COUNT, 3).
-define(ACTIVITY_SLEEP, 500).
-define(URGENT_SIGNAL, "urgent_signal").
-define(EVICTION_HSB, 500_000).

execute(_Context, [[ActivityPayloadSize, EvictionStrategy]]) when
    is_integer(ActivityPayloadSize), is_binary(EvictionStrategy)
->
    start_execution(eviction_handler, {binary_to_existing_atom(EvictionStrategy), 1}),
    ActivityPayload = binary:copy(~"X", ActivityPayloadSize),
    run_blocking_activities(ActivityPayload, ?ACTIVITY_COUNT),
    AwaitResult = temporal_sdk_workflow:await(
        {signal_request, ?URGENT_SIGNAL},
        [evict, {timeout, {2, second}}]
    ),
    run_blocking_activities(ActivityPayload, ?ACTIVITY_COUNT),
    case AwaitResult of
        {ok, #{state := requested}} -> admit_signal(?URGENT_SIGNAL);
        _ -> ok
    end,
    run_blocking_activities(ActivityPayload, ?ACTIVITY_COUNT),
    ok.

run_blocking_activities(Payload, ActivityCount) ->
    [
        #{state := completed} = start_activity(echo_activity, [Payload, ?ACTIVITY_SLEEP], [wait])
     || _ <- lists:seq(1, ActivityCount)
    ].

eviction_handler(#{}, {EvictionStrategy, EventId}) ->
    case await({event, {EventId, '_', '_', '_'}}) of
        {noevent, noevent} ->
            ok;
        {ok, #{event_id := EventId, attributes := #{history_size_bytes := HSB}}} ->
            maybe_evict(EvictionStrategy, HSB),
            eviction_handler(#{}, {EvictionStrategy, EventId + 1});
        {ok, #{event_id := EventId}} ->
            eviction_handler(#{}, {EvictionStrategy, EventId + 1})
    end.

maybe_evict(always, _HistorySizeBytes) -> do_evict();
maybe_evict(custom, HistorySizeBytes) when HistorySizeBytes > ?EVICTION_HSB -> do_evict();
maybe_evict(_EvictionStrategy, _HistorySizeBytes) -> ok.

do_evict() ->
    case is_awaited({signal, ?URGENT_SIGNAL}) of
        {true, #{state := requested}} -> ok;
        _ -> evict_workflow()
    end.

File: lib/eviction_parallel_handler/eviction_parallel_handler_workflow.ex
defmodule EvictionParallelHandler.Workflow do
  use TemporalSdk.Workflow
  @moduledoc false

  @activity_count 3
  @activity_sleep 500
  @urgent_signal "urgent_signal"
  @eviction_hsb 500_000

  @impl true
  def execute(_context, [[activity_payload_size, handle_eviction_strategy]])
      when is_integer(activity_payload_size) and is_binary(handle_eviction_strategy) do
    start_execution(:eviction_handler, {String.to_existing_atom(handle_eviction_strategy), 1})
    activity_payload = String.duplicate("X", activity_payload_size)
    run_blocking_activities(activity_payload, @activity_count)

    await_result =
      TemporalSdk.Workflow.await({:signal_request, @urgent_signal}, [
        :evict,
        timeout: {2, :second}
      ])

    run_blocking_activities(activity_payload, @activity_count)

    case await_result do
      {:ok, %{state: :requested}} -> admit_signal(@urgent_signal)
      _ -> :ok
    end

    run_blocking_activities(activity_payload, @activity_count)
    :ok
  end

  def run_blocking_activities(payload, activity_count) do
    for _ <- 1..activity_count do
      %{state: :completed} = start_activity(EchoActivity, [payload, @activity_sleep], [:wait])
    end
  end

  def eviction_handler(%{}, {eviction_strategy, event_id}) do
    case await({:event, {event_id, :_, :_, :_}}) do
      {:noevent, :noevent} ->
        :ok

      {:ok, %{event_id: event_id, attributes: %{history_size_bytes: hsb}}} ->
        maybe_evict(eviction_strategy, hsb)
        eviction_handler(%{}, {eviction_strategy, event_id + 1})

      {:ok, %{event_id: event_id}} ->
        eviction_handler(%{}, {eviction_strategy, event_id + 1})
    end
  end

  def maybe_evict(:always, _history_size_bytes), do: do_evict()

  def maybe_evict(:custom, history_size_bytes) when history_size_bytes > @eviction_hsb,
    do: do_evict()

  def maybe_evict(_eviction_strategy, _history_size_bytes), do: :ok

  def do_evict() do
    case is_awaited({:signal, @urgent_signal}) do
      {true, %{state: :requested}} -> :ok
      _ -> evict_workflow()
    end
  end
end
